runtime/gc_vlhgc/IncrementalGenerationalGC.cpp

```diff
@@ -1179,98 +1281,16 @@ MM_IncrementalGenerationalGC::attemptHeapResize(MM_EnvironmentVLHGC *env, MM_All
-		 * so that we can avoid this kind of stats-passing mess.  
-		 */
-		double regionConsumptionRate = _schedulingDelegate.getTotalRegionConsumptionRate();
-		double avgSurvivorRegions = _schedulingDelegate.getAverageSurvivorSetRegionCount();
-		double avgCopyForwardRate = _schedulingDelegate.getAverageCopyForwardRate();
-		U_64 scanTimeCostPerGMP = _schedulingDelegate.getScanTimeCostPerGMP(env);
-
-		double optimalEmptinessRegionThreshold = _reclaimDelegate.calculateOptimalEmptinessRegionThreshold(env, regionConsumptionRate, avgSurvivorRegions, avgCopyForwardRate, scanTimeCostPerGMP);
-		_schedulingDelegate.setAutomaticDefragmentEmptinessThreshold(optimalEmptinessRegionThreshold);
-	}
-
-	/* Determine if there are enough regions available to attempt a copy-forward collection.
-	 * Note that this check is done after we sweep, since that might have recovered enough 
-	 * regions to make copy-forward feasible.
-	 */ 
-	if (env->_cycleState->_shouldRunCopyForward) {
-		MM_GlobalAllocationManagerTarok *allocationmanager = (MM_GlobalAllocationManagerTarok *)_extensions->globalAllocationManager; 
-		/* We need at least one region per context for surviving Eden objects. Older objects can move into tail-fill regions. */
-		/* (We could require more than one region, but experimental evidence shows little value in doing so.) */ 
-		UDATA minimumRegionsForCopyForward = allocationmanager->getManagedAllocationContextCount();
-		UDATA freeRegions = allocationmanager->getFreeRegionCount();
-		if (freeRegions < minimumRegionsForCopyForward) {
-			env->_cycleState->_shouldRunCopyForward = false;
-			env->_cycleState->_reasonForMarkCompactPGC = MM_CycleState::reason_insufficient_free_space;
-		}
-	}
-
-	partialGarbageCollectUsingCopyForward(env, allocDescription);
-
-	env->_cycleState->_workPackets = NULL;
-	env->_cycleState->_markMap = NULL;
-
-	if (attemptHeapResize(env, allocDescription)) {
-		/* Check was it successful contraction */
-		if (env->_cycleState->_activeSubSpace->wasContractedThisGC(_extensions->globalVLHGCStats.gcCount)) {
-			_interRegionRememberedSet->setShouldFlushBuffersForDecommitedRegions();
-		}
-	}
-
-	env->_cycleState->_externalCycleState = NULL;
-
-	incrementRegionAges(env, _taxationThreshold, true);
-
-	reportGCCycleFinalIncrementEnding(env);
-	reportGCIncrementEnd(env);
-	reportPGCEnd(env);
-	reportGCCycleEnd(env);
-
-	/* Reset amount allocated for next PGC */
-	_allocatedSinceLastPGC = 0;
-
-	_extensions->allocationStats.clear();
-}
-
 void
-MM_IncrementalGenerationalGC::partialGarbageCollectUsingCopyForward(MM_EnvironmentVLHGC *env, MM_AllocateDescription *allocDescription)
+MM_IncrementalGenerationalGC::preProcessPGCUsingCopyForward(MM_EnvironmentVLHGC *env, MM_AllocateDescription *allocDescription)
 {
-	Trc_MM_IncrementalGenerationalGC_partialGarbageCollectUsingCopyForward_Entry(env->getLanguageVMThread());
-	
 	PORT_ACCESS_FROM_ENVIRONMENT(env);
+	MM_CycleState *cycleState = env->_cycleState;
```

Comment:
why not of type MM_CycleStateVLHGC, cast once here and avoid casting later at use sites?

Reply:
True, should have noticed that. Updating it.